IncludeToolbox/IncludeFormatter/IncludeFormatter.cs

```diff
@@ -86,24 +87,86 @@ public static void FormatSlashes(IncludeLineInfo[] lines, FormatterOptionsPage.S
             }
         }
 
-        public static void SortIncludes(IncludeLineInfo[] lines, FormatterOptionsPage.TypeSorting typeSorting, bool regexIncludeDelimiter, string[] precedenceRegexes, string documentName)
+        public const string CurrentFileNameKey = "$(currentFilename)";
+
+        private static string[] FixupRegexes(string[] precedenceRegexes, string documentName)
+        {
+            string currentFilename = documentName.Substring(0, documentName.LastIndexOf('.'));
+
+            string[] regexes = new string[precedenceRegexes.Length];
+            for (int i = 0; i < precedenceRegexes.Length; ++i)
+            {
+                regexes[i] = precedenceRegexes[i].Replace(CurrentFileNameKey, currentFilename);
+            }
+            return regexes;
+        }
+
+        public static void SortIncludes(IncludeLineInfo[] lines, FormatterOptionsPage settings, string documentName)
         {
-            var comparer = new IncludeComparer(precedenceRegexes, documentName);
-            var sortedIncludes = lines.Where(x => x.LineType != IncludeLineInfo.Type.NoInclude).OrderBy(x => x.IncludeContentForRegex(regexIncludeDelimiter), comparer);
+            FormatterOptionsPage.TypeSorting typeSorting = settings.SortByType;
+            bool regexIncludeDelimiter = settings.RegexIncludeDelimiter;
+            bool blankAfterRegexGroupMatch = settings.BlankAfterRegexGroupMatch;
+            bool removeEmptyLines = settings.RemoveEmptyLines;
+
+            string[] precedenceRegexes = FixupRegexes(settings.PrecedenceRegexes, documentName);
+
+            // Select only valid include lines and sort them. They'll stay in this relative sorted
+            // order when rearranged by regex precedence groups.
+            var includeLines = lines
+                .Where(x => x.LineType != IncludeLineInfo.Type.NoInclude)
+                .OrderBy(x => x.IncludeContent);
```

Comment:
Sort the lines alphabetically first. They'll maintain this relative order throughout the next steps.